In [1]:
# House Price Detections 

In [43]:
# impoting all the dependencies Liberies.
import numpy as np
import pandas as pd 
import seaborn as sns
from sklearn.model_selection import train_test_split 
from sklearn.preprocessing import StandardScaler 
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error, mean_squared_error
import statsmodels.api as sm

# Importing the Data set

In [44]:
house_df=pd.read_csv("D:\\Data Sets\\kc_house_data.csv")

In [45]:
house_df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000 221900.00         3       1.00         1180   
1  6414100192  20141209T000000 538000.00         3       2.25         2570   
2  5631500400  20150225T000000 180000.00         2       1.00          770   
3  2487200875  20141209T000000 604000.00         4       3.00         1960   
4  1954400510  20150218T000000 510000.00         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650    1.00           0     0  ...      7        1180              0   
1      7242    2.00           0     0  ...      7        2170            400   
2     10000    1.00           0     0  ...      6         770              0   
3      5000    1.00           0     0  ...      7        1050            910   
4      8080    1.00           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode   lat    long  sqft_living15  sqft_lot15  
0      1955             0    98178 47.51 -122.26           1340        5650  
1      1951          1991    98125 47.72 -122.32           1690        7639  
2      1933             0    98028 47.74 -122.23           2720        8062  
3      1965             0    98136 47.52 -122.39           1360        5000  
4      1987             0    98074 47.62 -122.05           1800        7503  

[5 rows x 21 columns]

# Performing the Exploratory Data Analysis 

In [46]:
house_df.shape

(21613, 21)

In [47]:
house_df.describe()

id      price  bedrooms  bathrooms  sqft_living   sqft_lot  \
count      21613.00   21613.00  21613.00   21613.00     21613.00   21613.00   
mean  4580301520.86  540088.14      3.37       2.11      2079.90   15106.97   
std   2876565571.31  367127.20      0.93       0.77       918.44   41420.51   
min      1000102.00   75000.00      0.00       0.00       290.00     520.00   
25%   2123049194.00  321950.00      3.00       1.75      1427.00    5040.00   
50%   3904930410.00  450000.00      3.00       2.25      1910.00    7618.00   
75%   7308900445.00  645000.00      4.00       2.50      2550.00   10688.00   
max   9900000190.00 7700000.00     33.00       8.00     13540.00 1651359.00   

        floors  waterfront     view  condition    grade  sqft_above  \
count 21613.00    21613.00 21613.00   21613.00 21613.00    21613.00   
mean      1.49        0.01     0.23       3.41     7.66     1788.39   
std       0.54        0.09     0.77       0.65     1.18      828.09   
min       1.00        0.00     0.00       1.00     1.00      290.00   
25%       1.00        0.00     0.00       3.00     7.00     1190.00   
50%       1.50        0.00     0.00       3.00     7.00     1560.00   
75%       2.00        0.00     0.00       4.00     8.00     2210.00   
max       3.50        1.00     4.00       5.00    13.00     9410.00   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
count       21613.00  21613.00      21613.00 21613.00 21613.00 21613.00   
mean          291.51   1971.01         84.40 98077.94    47.56  -122.21   
std           442.58     29.37        401.68    53.51     0.14     0.14   
min             0.00   1900.00          0.00 98001.00    47.16  -122.52   
25%             0.00   1951.00          0.00 98033.00    47.47  -122.33   
50%             0.00   1975.00          0.00 98065.00    47.57  -122.23   
75%           560.00   1997.00          0.00 98118.00    47.68  -122.12   
max          4820.00   2015.00       2015.00 98199.00    47.78  -121.31   

       sqft_living15  sqft_lot15  
count       21613.00    21613.00  
mean         1986.55    12768.46  
std           685.39    27304.18  
min           399.00      651.00  
25%          1490.00     5100.00  
50%          1840.00     7620.00  
75%          2360.00    10083.00  
max          6210.00   871200.00

In [48]:
# As our data is showing values is scientific notations need to fix it 
pd.options.display.float_format = '{:.2f}'.format

In [49]:
house_df.describe()

id      price  bedrooms  bathrooms  sqft_living   sqft_lot  \
count      21613.00   21613.00  21613.00   21613.00     21613.00   21613.00   
mean  4580301520.86  540088.14      3.37       2.11      2079.90   15106.97   
std   2876565571.31  367127.20      0.93       0.77       918.44   41420.51   
min      1000102.00   75000.00      0.00       0.00       290.00     520.00   
25%   2123049194.00  321950.00      3.00       1.75      1427.00    5040.00   
50%   3904930410.00  450000.00      3.00       2.25      1910.00    7618.00   
75%   7308900445.00  645000.00      4.00       2.50      2550.00   10688.00   
max   9900000190.00 7700000.00     33.00       8.00     13540.00 1651359.00   

        floors  waterfront     view  condition    grade  sqft_above  \
count 21613.00    21613.00 21613.00   21613.00 21613.00    21613.00   
mean      1.49        0.01     0.23       3.41     7.66     1788.39   
std       0.54        0.09     0.77       0.65     1.18      828.09   
min       1.00        0.00     0.00       1.00     1.00      290.00   
25%       1.00        0.00     0.00       3.00     7.00     1190.00   
50%       1.50        0.00     0.00       3.00     7.00     1560.00   
75%       2.00        0.00     0.00       4.00     8.00     2210.00   
max       3.50        1.00     4.00       5.00    13.00     9410.00   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
count       21613.00  21613.00      21613.00 21613.00 21613.00 21613.00   
mean          291.51   1971.01         84.40 98077.94    47.56  -122.21   
std           442.58     29.37        401.68    53.51     0.14     0.14   
min             0.00   1900.00          0.00 98001.00    47.16  -122.52   
25%             0.00   1951.00          0.00 98033.00    47.47  -122.33   
50%             0.00   1975.00          0.00 98065.00    47.57  -122.23   
75%           560.00   1997.00          0.00 98118.00    47.68  -122.12   
max          4820.00   2015.00       2015.00 98199.00    47.78  -121.31   

       sqft_living15  sqft_lot15  
count       21613.00    21613.00  
mean         1986.55    12768.46  
std           685.39    27304.18  
min           399.00      651.00  
25%          1490.00     5100.00  
50%          1840.00     7620.00  
75%          2360.00    10083.00  
max          6210.00   871200.00

In [50]:
house_df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

* After analyzing the Data Set we found no null values.

* Droping ID, Date, Lat, Long column as they are non affecting variables to the price 

In [51]:
house_df = house_df.drop(["id", "date","lat","long"], axis=1)

In [52]:
house_df.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0 221900.00         3       1.00         1180      5650    1.00           0   
1 538000.00         3       2.25         2570      7242    2.00           0   
2 180000.00         2       1.00          770     10000    1.00           0   
3 604000.00         4       3.00         1960      5000    1.00           0   
4 510000.00         3       2.00         1680      8080    1.00           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0     0          3      7        1180              0      1955             0   
1     0          3      7        2170            400      1951          1991   
2     0          3      6         770              0      1933             0   
3     0          5      7        1050            910      1965             0   
4     0          3      8        1680              0      1987             0   

   zipcode  sqft_living15  sqft_lot15  
0    98178           1340        5650  
1    98125           1690        7639  
2    98028           2720        8062  
3    98136           1360        5000  
4    98074           1800        7503

# Splitting the data set into train and test data 

In [53]:
house_df.columns

Index(['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'sqft_living15',
       'sqft_lot15'],
      dtype='object')

In [54]:
# Provided feature variables and target variable
feature_columns = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
                   'waterfront', 'view', 'condition', 'grade', 'sqft_above',
                   'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'sqft_living15',
                   'sqft_lot15']
target_column = 'price'

df = pd.DataFrame(house_df, columns=feature_columns + [target_column])

X = df[feature_columns]
y = df[target_column]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20,random_state=None)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (17290, 16)
X_test shape: (4323, 16)
y_train shape: (17290,)
y_test shape: (4323,)


In [55]:
X_train.head()

bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  \
5307          4       3.25         4750     62365    2.00           0     0   
6297          3       2.75         3940      9671    1.00           0     4   
21331         4       2.50         2623      7184    2.00           0     0   
6513          3       2.50         2110      8164    2.00           0     0   
12614         4       2.50         3360     15091    2.00           0     0   

       condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
5307           3     11        4750              0      1988             0   
6297           5      9        2140           1800      1967             0   
21331          3      8        2623              0      2012             0   
6513           3      9        2110              0      1990             0   
12614          3      9        3360              0      1997             0   

       zipcode  sqft_living15  sqft_lot15  
5307     98052           3300       31866  
6297     98006           3390        9360  
21331    98092           2010        4939  
6513     98056           2390        7499  
12614    98052           1930        9936

In [56]:
y_train.head()

5307    1122500.00
6297    1450000.00
21331    333000.00
6513     535000.00
12614    740000.00
Name: price, dtype: float64

# Normalization the data set 

In [57]:
std=StandardScaler()
X_train_std = std.fit_transform(X_train)
X_test_std = std.transform(X_test)

# Modelling the data set

In [58]:
model = LinearRegression()

In [59]:
model.fit(X_train_std,y_train)

LinearRegression()

In [60]:
y_predict = model.predict(X_test_std)

In [61]:
y_predict

array([370780.45936254, 582403.12770586, 500578.81395663, ...,
       728772.63313314, 401690.35740958, 303199.34798547])

In [62]:
y_test

5635    364000.00
18618   355000.00
19615   338000.00
13471   745000.00
13794   350000.00
           ...   
7211    165000.00
6325    325000.00
21374   705000.00
19610   500000.00
20734   330675.00
Name: price, Length: 4323, dtype: float64

# Calculation RMSE, MSE and R-squared score

In [63]:
r2 = r2_score(y_test, y_predict)
mse = mean_squared_error(y_test, y_predict)
rmse = mean_squared_error(y_test, y_predict, squared=False)

In [64]:
print("Mean Squared Error (MSE):", mse)
print("Root Mean Squared Error (RMSE):", rmse)
print("R-squared score:", r2)

Mean Squared Error (MSE): 49226924632.362
Root Mean Squared Error (RMSE): 221871.4146355091
R-squared score: 0.6464141875096092


## Model Performance Insights:

* Mean Squared Error (MSE):49226924632.362

This shows how far, on average, our predictions are from the actual values.
A higher MSE means our predictions have bigger errors.
In this case, the model's predictions have quite a bit of variation from the actual values.

* Root Mean Squared Error (RMSE): 221871.4146355091

This is like the "average" size of our prediction errors.
A higher RMSE means our predictions are, on average, off by a larger amount.
The given RMSE of 214352.73 means our predictions are off by around 214,352.73 units on average.

* R-squared score: 0.6464141875096092

This tells us how well our model fits the data compared to a simple average.
A higher R-squared score means our model explains more of the variability in the data.
An R-squared of 0.67 means our model explains about 67% of the variability, which is pretty good but not perfect.

# Inprovements need to be done

* Decreasing the test_size of the data

* Again Performing the modelling with text size 15.

In [68]:
feature_columns = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
                   'waterfront', 'view', 'condition', 'grade', 'sqft_above',
                   'sqft_basement', 'yr_built', 'yr_renovated', 'sqft_living15',
                   'zipcode','sqft_lot15']
target_column = 'price'

df = pd.DataFrame(house_df, columns=feature_columns + [target_column])

X = df[feature_columns]
y = df[target_column]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15,random_state=None)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (18371, 16)
X_test shape: (3242, 16)
y_train shape: (18371,)
y_test shape: (3242,)


In [69]:
X_train.head()

bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  \
21424         5       3.25         4100      7578    2.00           0     2   
10312         3       1.00         1480      6550    1.50           0     0   
11036         2       1.75         2220     33000    1.00           0     0   
4738          3       2.25         1650      2958    2.00           0     0   
11961         3       2.50         1990      8325    2.00           0     0   

       condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
21424          3     10        4100              0      2011             0   
10312          4      6        1480              0      1925             0   
11036          4      8        2220              0      1970             0   
4738           3      7        1650              0      1985             0   
11961          4      8        1310            680      1997             0   

       sqft_living15  zipcode  sqft_lot15  
21424           3710    98075        8156  
10312           1200    98126        6550  
11036           2220    98072       33000  
4738            1510    98031        3788  
11961           2450    98034       10606

In [70]:
X_train_std = std.fit_transform(X_train)
X_test_std = std.transform(X_test)

In [71]:
new_model= LinearRegression()

In [72]:
new_model.fit(X_train_std,y_train)

LinearRegression()

In [73]:
y_predict = new_model.predict(X_test_std)

In [74]:
y_predict

array([941998.49140192, 495226.93497214, 631846.14834196, ...,
       626966.35712099, 886394.07679728, 686226.25662878])

In [75]:
y_test

15734    812000.00
6118     325000.00
14127    868500.00
6298     490000.00
6865     550000.00
           ...    
15745    912000.00
15760    182200.00
20675    680000.00
3439    1280600.00
12061    531000.00
Name: price, Length: 3242, dtype: float64

In [76]:
new_r2 = r2_score(y_test, y_predict)
new_mse = mean_squared_error(y_test, y_predict)
new_rmse = mean_squared_error(y_test, y_predict, squared=False)
print("New Mean Squared Error (MSE):", new_mse)
print("New Root Mean Squared Error (RMSE):", new_rmse)
print("New R-squared score:", new_r2)

New Mean Squared Error (MSE): 41701578080.36699
New Root Mean Squared Error (RMSE): 204209.64247646826
New R-squared score: 0.6670180142990337


# Original Model Performance:

* Mean Squared Error (MSE): 49,226,924,632.362
* Root Mean Squared Error (RMSE): 221,871.4146355091
* R-squared Score: 0.6464
* The original model had a high MSE and RMSE, indicating that the predicted values were on average quite far from the actual values. The R-squared score of 0.6464 suggests that around 64.64% of the variability in the dependent variable was explained by the model. While this indicates some level of predictability, there's room for improvement.

# Improved Model Performance (After Decreasing Test Size):

* New Mean Squared Error (MSE): 41,701,578,080.36699
* New Root Mean Squared Error (RMSE): 204,209.64247646826
* New R-squared Score: 0.6670


* After decreasing the test size, the model's performance improved. The new MSE and RMSE values are lower, which means the model's predictions are closer to the actual values compared to the original model. The new R-squared score of 0.6670 indicates that the model now explains approximately 66.70% of the variability in the dependent variable, showing a notable improvement in predictive power.

## Impact of Decreasing Test Size:

Decreasing the test size likely led to a larger training set, which might have allowed the model to capture more intricate patterns in the data and make better predictions.
A larger training set can contribute to reducing overfitting, as the model learns from more diverse examples, leading to better generalization on the test set.

* Model Applicability:

The improved R-squared score suggests that the model's predictions are better aligned with the actual data. However, it's important to remember that an R-squared score of 0.6670 still indicates that there is unexplained variability in the data.

* Further Steps:

While the model has improved, there might still be opportunities for enhancing its performance. Consider exploring different algorithms, feature engineering, or hyperparameter tuning to achieve even better results.
It's essential to validate the model's performance on unseen data to ensure that the improvements are consistent across different datasets.